File: advent.lua
#! /bin/env lua
--[[-%tabs=4-----------------------------------------------------------------
|	                                                                    	|
|	Module:		advent.lua                                               	|
|	Function:	Load data for Advent of Code puzzle                        	|
|	Created:	16:04:10  15 Dec  2018                                  	|
|	Author:		Andrew Cannon <ajc@gmx.net>                                	|
|	                                                                    	|
|	Copyright(c) 2018-2019 Andrew Cannon                                  	|
|	Licensed under the terms of the MIT License                            	|
|	                                                                    	|
]]---------------------------------------------------------------------------
local advent = {}
if _VERSION:match"Lua 5%.[12]" then
	module("advent",package.seeall)
	advent = _G.advent
end


local function loadData(file, block)
	local fd = file == '-' and io.stdin or io.open(file, "r")
	local start = true
	local active = not block
	local out = {}
	local blank = "^%s*$"
	local sentinel
	local havesentinel
	local lnum = 1
	
--	print(file)
	
	if not fd then return nil, "data file not found" end
	
--	print(block)
	
	local getline = fd:lines()
	for line in getline do
		lnum = lnum + 1
		if line:match"^%s*$" and (not sentinel or havesentinel) then
--			print("eop at "..lnum)
			if active then
				break
			else
				start = true
			end
		elseif start and line:match"%S+" == block then
			sentinel = line:match"%S+%s+(%S+)"
			
--			print("found block " .. block)
			if sentinel then
				sentinel = "^" .. sentinel:gsub("%^%$%(%)%%%.%[%]%*%+%-%?", "%1") .. "%s*$"
--				print("sentinel="..sentinel)
			end
			active = true
		elseif active then
			havesentinel = sentinel and line:match(sentinel)
			table.insert(out, line)
		elseif start then
			start = false
		end
	end
	if file ~= '-' then
		fd:close()
	end	
	
	if not active then return nil, "data not found" end
	
	return table.concat(out, "\n") .. "\n", out
end

local function getData(x)
	local name = arg[0]:match"advent%d%d%d%d"
--	print(name)
	if x then
		if x == true or tonumber(x) then
			name = name .. "example" .. (x == true and "" or tostring(x))
		else
			return loadData(x)	-- x is filename or "-"
		end
	end
	return loadData(name:sub(1,8) .. ".dat", name)
end

advent.getData = getData

return advent

File: Advent.hs
{-
--[[-%tabs=6-------------------------------------------------------------------
|	                                                                    	|
|	Module:	advent.hs                                               	|
|	Function:	Load data for Advent of Code puzzle                        	|
|	Created:	16:04:10  15 Dec  2018                                  	|
|	Author:	Andrew Cannon <ajc@gmx.net>                                	|
|	                                                                    	|
|	Copyright(c) 2018-2019 Andrew Cannon                                  	|
|	Licensed under the terms of the MIT License                            	|
|	                                                                    	|
-------------------------------------------------------------------------------
-}

module Advent where

import System.Environment
import Data.List

data Mode = Normal | Example (Maybe Int) | File String

getData :: Mode -> IO [[String]]
getData mode = do
	progName <- getProgName
	let file = case mode of
		File f -> f
		otherwise -> take 8 progName ++ ".dat"
	content <- readFile file
	let	key = case mode of
			Normal -> progName
			Example Nothing -> progName ++ "example"
			Example (Just n) -> progName ++ "example" ++ show n
			File _ -> ""
		flines = lines content
		in case mode of
			File _ -> return [ flines ]
			otherwise ->
				let	(_, flines') = break (key `isPrefixOf`) flines
					sentinel = unwords $ tail $ words $ head flines'
					flines'' = tail flines'
					indata x = (not $ null x) && (not $ null $ head x)
					sep c s =
						let (s', s'') = break (c ==) s
						in case s'' of
						[] -> [s']
						otherwise -> s':sep c (tail s'')
				in return $
					if null flines' then error "*** data not found ***"
					else case sentinel of
					"" -> [ head $ sep "" flines'' ]
					s -> takeWhile indata $ sep s flines''
